### PR TITLE
Support pulling system imagelist cache from bundled charts

### PR DESCRIPTION
**Pull request by aiyengar2**
Currently, **for system-charts only**, we use the same scripts we use to generate the rancher-images.txt in order to produce a ConfigMap called `system-library-catalog-image-list` in the `cattle-system` namespace.

This ConfigMap contains a list of images that were extracted from system-charts stored in `management-state/catalog-cache/{HASH-OF-CATALOG}`.

However, for air-gapped setups, the system-charts do not get pulled into this directory. Instead, they only exist as packaged in the Rancher [Dockerfile](https://github.com/rancher/rancher/blob/522c7764153e4cfee6161895ae44efaa59d611d9/package/Dockerfile#L53) under `/var/lib/rancher-data/local-catalogs/system-library`. So the creation of the ConfigMap fails, triggering the 500 error seen on the ticket.

The fix pushed out in this PR modifies the functions that are used to create this ConfigMap to strip it into an exported helper function. This helper function is then added into the `doAddCatalogs` step of a Rancher startup (which is triggered every time the MCM controller starts).

Related Issue: https://github.com/rancher/rancher/issues/34172

### CODE REVIEW

File: pkg/catalog/manager/catalog_sync.go
Comment (aiyengar2) [inline pkg/catalog/manager/catalog_sync.go:76]: The original code utilized both a Lister (for Gets) and a Interface (for Create / Update) instead of just doing all calls via the Interface.

I debated making all calls go through the Interface instead to keep the function signature simple but chose to stick with this approach since I did not want to have to reach out to the k8s API in order to check if the ConfigMap exists; instead, looking through Rancher's cache of ConfigMaps via the Lister is sufficient and more efficient IMO.

Feel free to comment if you think I should drop the Lister in favor of just doing calls via the Interface!

File: pkg/data/management/catalog_data.go
Comment (aiyengar2) [inline pkg/data/management/catalog_data.go:210]: For this invocation of `CreateOrUpdateSystemCatalogImageCache`, I am specifying that we always do a forceUpdate since, if a user were to change the contents of `/var/lib/rancher-data/local-catalogs/system-library`, they should be able to update the contents of the `/v3/kontainerdriver/rancher-images` simply by restarting Rancher. 

Without this, there would be no way of supporting upgrading the system-charts and seeing the new images pop up in air-gapped setups.

File: pkg/image/resolve.go
Comment (aiyengar2) [inline pkg/image/resolve.go:352]: This function is no longer used by Rancher, but it is a public facing API. Should I leave this function behind in case someone is using Rancher as a library? Not sure what our policy is with deprecating such exported functions.

File: pkg/catalog/utils/utils.go
Comment (aiyengar2) [inline pkg/catalog/utils/utils.go:96]: I didn't need to add all these switch cases, but I did for good measure in case someone wants to use this helper function in the future 😄

File: pkg/image/resolve.go
Comment (PennyScissors) [reply pkg/image/resolve.go:352]: If we aren't using it, I would prefer to remove it since we already export GetImages() and this package can use some cleanup.

Comment (aiyengar2) [reply pkg/image/resolve.go:352]: cc: @kinarashah what do you think?

Comment (kinarashah) [reply pkg/image/resolve.go:352]: `CreateCatalogImageListConfigMap` was introduced very recently by https://github.com/rancher/rancher/pull/33151 for 2.6. I don't think this has been released yet, was this elsewhere?

Comment (aiyengar2) [reply pkg/image/resolve.go:352]: Ah yes I think you are right; we should be safe to remove it then. Thanks @kinarashah @PennyScissors!

**Comment by aiyengar2** [comment]
Need to create backport PR for release/v2.6 once complete, but will wait on approvals before setting up PR.

File: pkg/catalog/manager/catalog_sync.go
Comment (PennyScissors) [inline pkg/catalog/manager/catalog_sync.go:123]: nit: remove this comment, is unnecessary

**Review by PennyScissors — DISMISSED** [review]
lgtm, just one nit

File: pkg/catalog/manager/catalog_sync.go
Comment (aiyengar2) [reply pkg/catalog/manager/catalog_sync.go]: addressed

**Comment by aiyengar2** [comment]
addressed @PennyScissors's nit

File: pkg/catalog/manager/catalog_sync.go
Comment (kinarashah) [inline pkg/catalog/manager/catalog_sync.go:114]: what's the use case for updating if systemCatalogImageCacheCreated=true, only forceUpdate=true should suffice?

File: pkg/data/management/catalog_data.go
Comment (kinarashah) [reply pkg/data/management/catalog_data.go:210]: This context should also be added as a comment on L207 to explain why we're force updating the catalog cache.

Comment (kinarashah) [reply pkg/data/management/catalog_data.go:210]: PS: how does `system-library` content affect `kontainerdriver` images? we only cache images from chart in `AddImagesToImageListConfigMap`?

File: pkg/catalog/manager/catalog_sync.go
Comment (kinarashah) [reply pkg/catalog/manager/catalog_sync.go:76]: Using Lister for Gets and client for Create+Updates sounds good. If the logic is expected to always return object if it exists, we use Interface to Get if Lister returns NotFound but in this case can just use client to Create and no-op if AlreadyExists as you did. 👍🏼

File: pkg/catalog/manager/catalog_sync.go
Comment (aiyengar2) [reply pkg/catalog/manager/catalog_sync.go:114]: This was copied from the original function's logic in https://github.com/rancher/rancher/blob/fa481ddb14e3a3413ca5d75e709474bfb02a7eab/pkg/catalog/manager/catalog_sync.go#L98

I believe the edge case it catches is if the ConfigMap is created by another Rancher instance during a `Sync` on the catalog manager.

If a ConfigMap were to be created after this line:
https://github.com/rancher/rancher/blob/fa481ddb14e3a3413ca5d75e709474bfb02a7eab/pkg/catalog/manager/catalog_sync.go#L72

Then our Create call in the following line can return an `IsAlreadyExist`:
https://github.com/rancher/rancher/blob/fa481ddb14e3a3413ca5d75e709474bfb02a7eab/pkg/catalog/manager/catalog_sync.go#L86-L87

In this case, if we are doing a forceUpdate, then we should still upgrade the ConfigMap with the latest.

Comment (kinarashah) [reply pkg/catalog/manager/catalog_sync.go:114]: Yeah, this might help resolve this, I think the create through any sync would have already fetched the latest images but makes sense to leave it as is from old logic than remove it.

File: pkg/data/management/catalog_data.go
Comment (aiyengar2) [reply pkg/data/management/catalog_data.go:210]: Sure, I will add it as a comment.

Comment (aiyengar2) [reply pkg/data/management/catalog_data.go:210]: > PS: how does system-library content affect kontainerdriver images? we only cache images from chart in AddImagesToImageListConfigMap?

`kontainerdriver` images are composed of images from 3 sources, as shown in https://github.com/rancher/rancher/blob/fa481ddb14e3a3413ca5d75e709474bfb02a7eab/pkg/api/norman/customization/kontainerdriver/actionhandler.go#L183-L185

1. The agent image
2. RKESystemImages (from KDM)
3. The system-chart images

Prior to 2.6, the contents of the git repo for system-charts in the `management-state` directory on each Rancher container was the source of truth for the latest copy of system-charts. We would `GetImages` on demand when such a request came in.

However, to fix #31336, we made the new source of truth for system-chart images the `system-library` ConfigMap instead, which already contains the output of the `GetImages` command on the system-charts.

Therefore, `AddImagesToImageListConfigMap` gets called in one of two ways:
1. On startup (after my fix) to create the initial ConfigMap and update it based on the bundled contents
2. On update (only in non-bundled Rancher clusters), as a result of the catalog `manager.Sync` if it is detected that the catalog is not up-to-date.

**Review by kinarashah — APPROVED** [review]

**Review by PennyScissors — APPROVED** [review]

**Pull request merged**
